includes/admin/class-wc-rest-payments-orders-controller.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Class WC_REST_Payments_Orders_Controller
+ *
+ * @package WooCommerce\Payments\Admin
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+use WCPay\Logger;
+
+/**
+ * REST controller for order processing.
+ */
+class WC_REST_Payments_Orders_Controller extends WC_Payments_REST_Controller {
+
+	/**
+	 * Endpoint path.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'payments/orders';
+
+	/**
+	 * Instance of WC_Payment_Gateway_WCPay
+	 *
+	 * @var WC_Payment_Gateway_WCPay
+	 */
+	private $gateway;
+
+	/**
+	 * WC_Payments_REST_Controller constructor.
+	 *
+	 * @param WC_Payments_API_Client   $api_client - WooCommerce Payments API client.
+	 * @param WC_Payment_Gateway_WCPay $gateway - WooCommerce Payments payment gateway.
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payment_Gateway_WCPay $gateway ) {
+		parent::__construct( $api_client );
+		$this->gateway = $gateway;
+	}
+
+	/**
+	 * Configure REST API routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/(?P<order_id>\w+)/capture_terminal_payment',
+			[
+				'methods'             => WP_REST_Server::CREATABLE,
+				'callback'            => [ $this, 'capture_terminal_payment' ],
+				'permission_callback' => [ $this, 'check_permission' ],
+				'args'                => [
+					'payment_intent_id' => [
+						'required' => true,
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Given an intent ID and an order ID, add the intent ID to the order and capture it.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 */
+	public function capture_terminal_payment( $request ) {
+		try {
+			$intent_id = $request['payment_intent_id'];
+			$order_id  = $request['order_id'];
+			$order     = wc_get_order( $order_id );
+			if ( ! $order ) {
+				return new WP_Error( 'wcpay_missing_order', __( 'Order not found', 'woocommerce-payments' ), [ 'status' => 404 ] );
+			}
+
+			$intent = $this->api_client->get_intent( $intent_id );
+
+			// Do not process intents that can't be captured.
+			if ( ! in_array( $intent->get_status(), [ 'processing', 'requires_capture' ], true ) ) {
+				return rest_ensure_response(
+					[
+						'status' => 'failed',
```

Comment:
Maybe this should return a WP_Error instead? Maybe with status 409?

Props @malinajirka

Reply:
Thank you so much for the list of possible responses @marcinbot!

Reply:
Ok I take that back, it seems that v2+ endpoints use "code". Sorry for the confusion.

Reply:
We have a handful of calls to Stripe, but they are wrapped by a client SDK they provide. The REST API requests we make ourselves are to WordPress, WooCommerce and WCPay endpoints and I think you're right - they all follow that same pattern.